lib/oli_web/controllers/cognito_controller.ex

```diff
@@ -0,0 +1,151 @@
+defmodule OliWeb.CognitoController do
+  use OliWeb, :controller
+  import Oli.HTTP
+  import Oli.Utils
+  import Ecto
+  import Ecto.Changeset
+  require Logger
+
+  alias Oli.Repo
+  alias Oli.Accounts
+  alias Oli.Groups.CommunityAccount
+  alias Oli.Institutions.SsoJwks
+
+  def launch(conn, params) do
+    try do
+      with {:ok, valid_params} <- validate_query_params(params) do
+        jwk = get_jwk(Map.get(params, "id_token"))
+        jwk_from_pem = JOSE.JWK.from_pem(jwk.pem)
+        case JOSE.JWT.verify_strict(jwk_from_pem, ["RS256"], valid_params.id_token) do
+          {true, verified_jwt, _} ->
+            case Accounts.insert_or_update_lms_user(%{
+                  sub: Map.get(verified_jwt.fields, "sub"),
+                  preferred_username: Map.get(verified_jwt.fields, "cognito:username"),
+                  email: Map.get(verified_jwt.fields, "email"),
+                  can_create_sections: true
+                }) do
+              {:ok, user} ->
+                  attrs = %{user_id: user.id, community_id: valid_params.community_id}
+
+                  %CommunityAccount{}
+                  |> CommunityAccount.changeset(attrs)
+                  |> Repo.insert()
+
+                  conn
+                  |> use_pow_config(:user)
+                  |> Pow.Plug.create(user)
+                  |> redirect(to: Routes.page_delivery_path(conn, :index, valid_params.course_section_id))
+            end
+
+          {false, _, _} ->
+            error_url = get_error_url(conn, valid_params)
+            conn
+            |> redirect(external: "#{error_url}?error=Unable to verify credentials")
+            |> halt()
+        end
+      else
+        {:error, changeset} ->
+          error_fields = changeset.errors
+          |> Keyword.keys()
+          |> Enum.join(", ")
+
+          error_url = get_error_url(conn, changeset.changes)
+          conn
+          |> redirect(external: "#{error_url}?error=#{error_fields} - missing or invalid")
+          |> halt()
+      end
+    rescue
+      e ->
+        uuid = Ecto.UUID.generate
+        Logger.error("#{uuid} - #{inspect(e)}")
+
+        error_url = get_error_url(conn, %{})
+        conn
+        |> redirect(external: "#{error_url}?error=Unknown error occurred - #{uuid}")
+    end
+  end
+
+  def get_jwk(id_token) do
+    jwt_header = id_token
+    |> Joken.peek_header()
+
+    jwt_payload = id_token
+    |> JOSE.JWT.peek_payload()
+
+    issuer = Map.get(jwt_payload.fields, "iss")
+
+    case Joken.peek_header(id_token) do
+      {:ok, result} ->
+        case Repo.get_by(SsoJwks, kid: Map.get(result, "kid")) do
+          nil ->
+            jwks = get_cognito_jwks(issuer)
+
+            sso_jwks = Enum.map(elem(jwks.keys, 1), fn(s) -> get_sso_jwks(s) end)
```

Comment:
I think this could work:

```elixir
sso_jwks =
  issuer
  |> get_cognito_jwks()
  |> Map.get(:keys)
  |> elem(1)
  |> Enum.map(& get_sso_jwks(&1))
```

or as we suggested via email

```elixir
%JOSE.JWK{keys: {_, keys_list}} = get_cognito_jwks(issuer)
sso_jwks = Enum.map(keys_list, & get_sso_jwks(&1))
```